packages/insured-bridge-relayer/src/Relayer.ts

```diff
@@ -1,19 +1,128 @@
+  // that processes relayableDeposits and calls respective relaying functions.
+  async checkForPendingDepositsAndRelay() {
+    this.logger.debug({ at: "Relayer", message: "Checking for pending deposits and relaying" });
+    const relayableDeposits: { [key: string]: any } = {};
+    for (const l1Token of this.whitelistedRelayL1Tokens) {
+      this.logger.debug({ at: "Relayer", message: "Checking relays for token", l1Token });
+      // TODO: consider limiting how far back we look in this call size.
+      const l2Deposits = this.l2Client.getAllDeposits();
+      relayableDeposits[l1Token] = l2Deposits.filter((l2Deposit) => !this.l1Client.hasL2DepositBeenRelayed(l2Deposit));
+      if (relayableDeposits[l1Token].length == 0) delete relayableDeposits[l1Token];
+    }
+    if (Object.keys(relayableDeposits).length == 0) {
+      this.logger.debug({ at: "Relayer", message: "No relayable deposits for any whitelisted tokens" });
+      return;
+    }
+    for (const l1Token of Object.keys(relayableDeposits)) {
+      this.logger.debug({ at: "Relayer", message: "Processing relayable deposits for L1Token", l1Token });
+      for (const relayableDeposit of relayableDeposits[l1Token]) {
+        const botL1TokenBalance = await this.fetchBotL1TokenBalance(l1Token);
+        const relayType = this.shouldRelay(relayableDeposit, botL1TokenBalance);
+        switch (relayType) {
+          case RelayType.No:
+            this.logger.info({
+              at: "Relayer",
+              message: "Not relaying deposit 😖",
+              l1Token,
+              relayableDeposit,
+              botL1TokenBalance,
+            });
+            break;
+          case RelayType.Slow:
+            this.logger.info({
+              at: "Relayer",
+              message: "Slow relaying deposit 🐌",
+              l1Token,
+              relayableDeposit,
+              botL1TokenBalance,
+            });
+            await this.slowRelay(relayableDeposit);
+            break;
+
+          case RelayType.Fast:
+            this.logger.info({
+              at: "Relayer",
+              message: "Fast relaying deposit 🏃‍♂️",
+              l1Token,
+              relayableDeposit,
+              botL1TokenBalance,
+            });
+            await this.fastRelay(relayableDeposit);
+            break;
+        }
+      }
+    }
+  }
+
+  private shouldRelay(deposit: Deposit, botTokenBalance: string): RelayType {
+    // shouldRelay(deposit) => RelayType { No, Slow, Fast }
+    // Checks bot balance.
+    // If bot balance is > transfer size, simulate the fast relay to determine the cost.
+    // If bot balance is > bond, simulate the slow relay to determine the cost.
+    // If the max profit across the two (or all that succeeded) is negative or 0, return No.
+    // Else return the one with the most profit.
+
+    // TODO: Add logic to decide if a relay action should happen
```

Comment:
Do we need to do any validation of the L2 deposit here? Or is that handled by the client?

Reply:
I think by it including in the client it is "valid". However, we do need to build out the Insured bridge price feed that will define the "validity" of a relay action. However, this is less important in this part of the infrastructure as the relayer will just simply do nothing if a bridging action is already relayed. Only the disputer bot needs to "validate" if something is valid or not and dispute accordingly.